pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,421 @@
+
+	Repo            policy.PolicyRepository
+	EndpointManager endpointmanager.EndpointManager
+	IPCache         *ipcache.IPCache
+	MonitorAgent    agent.Agent
+}
+
+type policyImporter struct {
+	log          *slog.Logger
+	repo         policy.PolicyRepository
+	epm          epmanager
+	ipc          ipcacher
+	monitorAgent agent.Agent
+
+	// prefixesByResources is the list of prefixes
+	// that belong to each resource. This is tracked separately
+	// so we can allocate and release prefixes as policy changes.
+	prefixesByResource map[ipcachetypes.ResourceID][]netip.Prefix
+
+	q chan *policytypes.PolicyUpdate
+}
+
+type ipcacher interface {
+	UpsertMetadataBatch(updates ...ipcache.MU) (revision uint64)
+	RemoveMetadataBatch(updates ...ipcache.MU) (revision uint64)
+	WaitForRevision(ctx context.Context, rev uint64) error
+}
+
+type epmanager interface {
+	UpdatePolicy(idsToRegen *set.Set[identity.NumericIdentity], fromRev, toRev uint64)
+}
+
+func newPolicyImporter(cfg policyImporterParams) PolicyImporter {
+	i := &policyImporter{
+		log:          cfg.Log,
+		repo:         cfg.Repo,
+		epm:          cfg.EndpointManager,
+		ipc:          cfg.IPCache,
+		monitorAgent: cfg.MonitorAgent,
+
+		q: make(chan *policytypes.PolicyUpdate, cfg.Config.PolicyQueueSize),
+
+		prefixesByResource: map[ipcachetypes.ResourceID][]netip.Prefix{},
+	}
+
+	buf := stream.Buffer(
+		stream.FromChannel(i.q),
+		int(cfg.Config.PolicyQueueSize), 10*time.Millisecond,
+		concat)
+
+	cfg.JobGroup.Add(job.Observer("policy-importer", i.processUpdates, buf))
+
+	return i
+}
+
+// ResourceIDAnonymous is the anonymous ipcache resource used as a placeholder
+// for policies that allocate CIDRs but do not have an owning resource.
+// (This is only used for policies created by the local API).
+const ResourceIDAnonymous = "policy/anonymous"
+
+func (i *policyImporter) UpdatePolicy(u *policytypes.PolicyUpdate) {
+	i.q <- u
+}
+
+func concat(buf []*policytypes.PolicyUpdate, in *policytypes.PolicyUpdate) []*policytypes.PolicyUpdate {
+	buf = append(buf, in)
+	return buf
+}
+
+// updatePrefixes determines the set of prefixes "owned" by a given resource and applies them
+// to the ipcache.
+// Write lock must be held.
+//
+// If the ipcache has started, it waits up to a configurable deadline for the prefixes
+// to be allocated.
+//
+// It returns the set of stale prefixes that should be deallocated after policy updates are complete.
+func (i *policyImporter) updatePrefixes(ctx context.Context, updates []*policytypes.PolicyUpdate) (toPrune map[ipcachetypes.ResourceID][]netip.Prefix) {
+	if i.ipc == nil {
+		return
+	}
+
+	// The set of all prefixes that belong to a resource.
+	toAllocate := map[ipcachetypes.ResourceID][]netip.Prefix{}
+	prefixSource := map[ipcachetypes.ResourceID]source.Source{}
+	toPrune = map[ipcachetypes.ResourceID][]netip.Prefix{}
+
+	// First, gather all new prefixes and index them by resource.
+	//
+	// For rules without an owning resource (k8s policy or file on disk),
```

Comment:
I think this comment is slightly confusing. I think you mean to say that rules without an owning resource come from the local REST API rather than from "k8s policy"?

Reply:
Right, the negative is unclear. I'll clarify.